jmxtrans-output/jmxtrans-output-elastic/src/main/java/com/googlecode/jmxtrans/model/output/elastic/ElasticWriter.java

```diff
@@ -0,0 +1,209 @@
+					log.debug("Insert into Elastic: Index: [{}] Type: [{}] Message: [{}]", indexName, TYPE_NAME, message);
+					Index index = new Index.Builder(message).index(indexName).type(TYPE_NAME).build();
+					jestClient.execute(index);
+				} else {
+					log.warn("Unable to submit non-numeric value to Elastic: [{}] from result [{}]", value, result);
+				}
+			}
+		}
+	}
+
+	private String createJsonMessage(Server server, Query query, List<String> typeNames, Result result, Entry<String, Object> values, Object value) throws IOException {
+
+		String keyString = getKeyString(server, query, result, values, typeNames, this.rootPrefix);
+
+		Closer closer = Closer.create();
+		try {
+			ByteArrayOutputStream out = closer.register(new ByteArrayOutputStream());
+			JsonGenerator generator = closer.register(jsonFactory.createGenerator(out, UTF8));
+			generator.writeStartObject();
+			generator.writeStringField("server", createAlias(server));
+			generator.writeStringField("metric", keyString);
+			generator.writeNumberField("value", Double.parseDouble(value.toString()));
+			generator.writeStringField("resultAlias", result.getKeyAlias());
+			generator.writeStringField("attributeName", result.getAttributeName());
+			generator.writeStringField("key", values.getKey());
+			generator.writeNumberField("timestamp", result.getEpoch());
+			generator.writeEndObject();
+			generator.close();
+			return out.toString("UTF-8");
+		} catch (Throwable t) {
+			throw closer.rethrow(t);
+		} finally {
+			closer.close();
+		}
+	}
+
+	private String createAlias(Server server) {
+		String alias;
+		if (server.getAlias() != null) {
+			alias = server.getAlias();
+		} else {
+			alias = server.getHost() + "_" + server.getPort();
+			alias = StringUtils.cleanupStr(alias);
+		}
+		return alias;
+	}
+
+	@VisibleForTesting
+	void setJestClient(JestClient jestClient) {
+		log.info("Note: using injected jestClient instead of default client: [{}]", jestClient);
+		this.jestClient = jestClient;
+	}
+
+	@Override
+	public void validateSetup(Server server, Query query) throws ValidationException {
+		// no validations
+	}
+
+	private static void createMappingIfNeeded(JestClient jestClient, String indexName, String typeName) throws IOException {
+		synchronized (CREATE_MAPPING_LOCK) {
+			IndicesExists indicesExists = new IndicesExists.Builder(indexName).build();
+			boolean indexExists = jestClient.execute(indicesExists).isSucceeded();
+
+			if (!indexExists) {
+
+				CreateIndex createIndex = new CreateIndex.Builder(indexName).build();
+				jestClient.execute(createIndex);
+
+				URL url = ElasticWriter.class.getResource("/elastic-mapping.json");
+				String mapping = Resources.toString(url, Charsets.UTF_8);
+
+				PutMapping putMapping = new PutMapping.Builder(indexName, typeName,mapping).build();
+
+				JestResult result = jestClient.execute(putMapping);
+				if (!result.isSucceeded()) {
+					log.warn("Failed to create mapping: {}", result.getErrorMessage());
```

Comment:
In case of error, we probably want to throw an exception to ensure that this writer is disabled. Unless this can be a transient error. So probably in case of communication error, we have a good chance of recovery, otherwise probably not so much ...